File: spring-web-java/src/main/java/spring/web/java/domain/item/repository/ItemCustomRepositoryImpl.java
package spring.web.java.domain.item.repository;

import static spring.web.java.domain.item.QItem.*;

import java.util.List;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageImpl;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.util.StringUtils;

import com.querydsl.core.types.dsl.BooleanExpression;
import com.querydsl.jpa.impl.JPAQueryFactory;

import jakarta.persistence.EntityManager;
import jakarta.persistence.TypedQuery;
import spring.web.java.domain.item.Item;

public class ItemCustomRepositoryImpl implements ItemCustomRepository {

	private final EntityManager entityManager;
	private final JPAQueryFactory queryFactory;

	public ItemCustomRepositoryImpl(EntityManager entityManager) {
		this.entityManager = entityManager;
		this.queryFactory = new JPAQueryFactory(entityManager);
	}

	@Override
	public Page<Item> findAllUsingJpql(Pageable pageable, String name) {
		String countJpql = """
			SELECT COUNT(i)
			FROM Item i
			WHERE 1 = 1
			""";
		String contentJpql = """
			SELECT i
			FROM Item i
			WHERE 1 = 1
			""";

		if (StringUtils.hasText(name)) {
			countJpql += " AND i.name LIKE CONCAT('%', :name, '%')";
			contentJpql += " AND i.name LIKE CONCAT('%', :name, '%')";
		}

		Sort sort = pageable.getSort();
		if (sort.isSorted()) {
			List<String> orderList = sort
				.map(order -> " i." + order.getProperty() + " " + order.getDirection())
				.toList();

			contentJpql += " ORDER BY " + String.join(",", orderList);
		}

		TypedQuery<Long> countQuery = entityManager.createQuery(countJpql, Long.class);
		TypedQuery<Item> contentQuery = entityManager.createQuery(contentJpql, Item.class)
			.setFirstResult((int)pageable.getOffset())
			.setMaxResults(pageable.getPageSize());

		if (StringUtils.hasText(name)) {
			countQuery.setParameter("name", name);
			contentQuery.setParameter("name", name);
		}

		return new PageImpl<>(contentQuery.getResultList(), pageable, countQuery.getSingleResult());
	}

	@Override
	public Page<Item> findAllUsingQueryDsl(Pageable pageable, String name) {
		int count = queryFactory
			.selectOne()
			.from(item)
			.where(likeName(name))
			.fetch()
			.size();

		List<Item> content = queryFactory
			.select(item)
			.from(item)
			.where(likeName(name))
			.limit(pageable.getPageSize())
			.offset(pageable.getOffset())
			.fetch();

		return new PageImpl<>(content, pageable, count);
	}

	private BooleanExpression likeName(String name) {
		if (!StringUtils.hasText(name)) {
			return null;
		}
		return item.name.like("%" + name + "%");
	}
}

File: spring-web-kotlin/src/main/kotlin/spring/web/kotlin/domain/item/repository/ItemCustomRepositoryImpl.kt
package spring.web.kotlin.domain.item.repository

import com.querydsl.core.types.dsl.BooleanExpression
import com.querydsl.jpa.impl.JPAQueryFactory
import jakarta.persistence.EntityManager
import jakarta.persistence.TypedQuery
import org.springframework.data.domain.Page
import org.springframework.data.domain.PageImpl
import org.springframework.data.domain.Pageable
import org.springframework.util.StringUtils
import spring.web.kotlin.domain.item.Item
import spring.web.kotlin.domain.item.QItem
import spring.web.kotlin.domain.item.QItem.item

class ItemCustomRepositoryImpl(
    private val entityManager: EntityManager,
    private val queryFactory: JPAQueryFactory = JPAQueryFactory(entityManager),
) : ItemCustomRepository {
    override fun findAllWithJpql(pageable: Pageable, name: String?): Page<Item> {
        var countJpql: String = """
            SELECT COUNT(i)
            FROM Item i
            WHERE 1 = 1
        """.trimIndent()

        var contentJpql: String = """
            SELECT i
            FROM Item i
            WHERE 1 = 1
        """.trimIndent()

        if (StringUtils.hasText(name)) {
            countJpql += " AND i.name LIKE CONCAT('%', :name, '%')"
            contentJpql += " AND i.name LIKE CONCAT('%', :name, '%')"
        }

        val sort = pageable.sort
        if (sort.isSorted) {
            val orderList = sort
                .map { order -> " i.${order.property} ${order.direction}" }
                .toList()

            contentJpql += " ORDER BY " + orderList.joinToString(",")
        }

        val countQuery: TypedQuery<Long> = entityManager.createQuery(countJpql, Long::class.java)
        val contentQuery: TypedQuery<Item> = entityManager.createQuery(contentJpql, Item::class.java)

        if (pageable.isPaged) {
            contentQuery.setFirstResult(pageable.offset.toInt()).setMaxResults(pageable.pageSize)
        }

        if (StringUtils.hasText(name)) {
            countQuery.setParameter("name", name)
            contentQuery.setParameter("name", name)
        }

        return PageImpl(contentQuery.resultList, pageable, countQuery.singleResult)
    }

    override fun findAllUsingQueryDsl(pageable: Pageable, name: String?): Page<Item> {
        val count = queryFactory
            .selectOne()
            .from(item)
            .where(likeName(name))
            .fetch()
            .size

        val content = queryFactory
            .select(item)
            .from(item)
            .where(likeName(name))
            .limit(pageable.pageSize.toLong())
            .offset(pageable.offset)
            .fetch()

        return PageImpl(content, pageable, count.toLong())
    }

    private fun likeName(name: String?) = if (name.isNullOrBlank()) null else item.name.like("%$name%")
}